### ISSUE TITLE
Segfault when trying to construct pv.PolyData and faces are specified incorrectly

### ISSUE THREAD

**Issue by sampotter** [pyvista/pyvista#1097]
I go to make a `pv.PolyData` from a triangle mesh specified by an Nx3 vertex and Mx3 face array. For example, I run:

```py
In [2]: import meshzoo
   ...: import numpy as np
   ...: import pyvista as pv
   ...: V, F = meshzoo.triangle(8)
   ...: V = V.T
   ...: WRONG_NUMBER = 1
   ...: pv.PolyData(
   ...:     V,
   ...:     np.concatenate([
   ...:         WRONG_NUMBER*np.ones((F.shape[0], 1), dtype=F.dtype),
   ...:         F
   ...:     ], axis=1)
   ...: )
```

and I get a segfault:

```
Out[2]: Fatal Python error: Segmentation fault

Thread 0x0000700006579000 (most recent call first):
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/history.py", line 764 in _writeout_input_cache
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/history.py", line 780 in writeout_cache
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/history.py", line 58 in needs_sqlite
  File "<decorator-gen-16>", line 2 in writeout_cache
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/history.py", line 834 in run
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/history.py", line 58 in needs_sqlite
  File "<decorator-gen-17>", line 2 in run
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 890 in _bootstrap

Current thread 0x000000010bdb2dc0 (most recent call first):
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/pyvista/core/common.py", line 742 in bounds
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/pyvista/core/common.py", line 857 in _get_attrs
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/pyvista/core/common.py", line 176 in head
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/pyvista/core/common.py", line 916 in __repr__
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/pyvista/core/pointset.py", line 192 in __repr__
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/lib/pretty.py", line 700 in _repr_pprint
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/lib/pretty.py", line 394 in pretty
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/formatters.py", line 702 in __call__
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/formatters.py", line 224 in catch_format_error
  File "<decorator-gen-3>", line 2 in __call__
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/formatters.py", line 180 in format
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/displayhook.py", line 151 in compute_format_data
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/displayhook.py", line 262 in __call__
  File "<ipython-input-2-467f16433641>", line 12 in <module>
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3418 in run_code
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3338 in run_ast_nodes
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3147 in run_cell_async
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/async_helpers.py", line 68 in _pseudo_sync_runner
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 2923 in _run_cell
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 2878 in run_cell
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/terminal/interactiveshell.py", line 555 in interact
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/terminal/interactiveshell.py", line 564 in mainloop
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/terminal/ipapp.py", line 356 in start
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/traitlets/config/application.py", line 845 in launch_instance
  File "/Users/sfp/LocalWorkspace/jmm/lib/python3.7/site-packages/IPython/__init__.py", line 126 in start_ipython
  File "/Users/sfp/LocalWorkspace/jmm/bin/ipython", line 8 in <module>
zsh: segmentation fault  ipython
```

Of course, I should have set `WRONG_NUMBER = 3` in the above example to avoid having this happen:

```py
In [7]: import meshzoo
   ...: import numpy as np
   ...: import pyvista as pv
   ...: V, F = meshzoo.triangle(8)
   ...: V = V.T
   ...: RIGHT_NUMBER = 3
   ...: pv.PolyData(
   ...:     V,
   ...:     np.concatenate([
   ...:         RIGHT_NUMBER*np.ones((F.shape[0], 1), dtype=F.dtype),
   ...:         F
   ...:     ], axis=1)
   ...: )
Out[7]:
PolyData (0x14300ede0)
  N Cells:	64
  N Points:	45
  X Bounds:	0.000e+00, 1.000e+00
  Y Bounds:	0.000e+00, 1.000e+00
  Z Bounds:	0.000e+00, 1.000e+00
  N Arrays:	0
```

For whatever reason, I make mistakes like this frequently, so PyVista often ends up crashing Python.

1. It would be good if PyVista threw an exception under these circumstances.
2. If it doesn't already have one, it would be very helpful if PyVista had a convenience method for building a `PolyData` from a vertex and face array pair, since this is such a common use case.

```
--------------------------------------------------------------------------------
  Date: Thu Jan 21 08:29:07 2021 EST

                OS : Darwin
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 16.0 GB
       Environment : IPython
        GPU Vendor : Intel Inc.
      GPU Renderer : Intel Iris Pro OpenGL Engine
       GPU Version : 4.1 INTEL-14.7.11

  Python 3.7.9 (default, Sep  6 2020, 13:20:25)  [Clang 11.0.3
  (clang-1103.0.32.62)]

           pyvista : 0.27.4
               vtk : 8.1.2
             numpy : 1.19.5
           imageio : 2.9.0
           appdirs : 1.4.4
            scooby : 0.5.6
            meshio : 4.3.8
        matplotlib : 3.3.3
         pyvistaqt : 0.3.0
             PyQt5 : 5.15.2
           IPython : 7.19.0
          colorcet : 1.0.0
             scipy : 1.6.0
--------------------------------------------------------------------------------
```

**Comment by akaszynski** [comment]
> It would be good if PyVista threw an exception under these circumstances.

Agreed.  VTK segfaults due to memory access issues, and it would be great to check the validity of the array prior to sending it to VTK.  However, the tradeoff would be that it would take longer to construct the `PolyData` as error checking it in numpy or pure python wouldn't be fast.  We could consider adding a `check_validity` parameter to the `PolyData` constructor.

> If it doesn't already have one, it would be very helpful if PyVista had a convenience method for building a PolyData from a vertex and face array pair, since this is such a common use case.

It doesn't, but we could add one fairly easily.  Would you mind submitting a PR that adds one to at `pyvista/utilities/helpers.py`?  Perhaps something like:

```python
def construct_mesh(points, faces):
   """Create a mesh given a square face connectivity array without padding.

    Examples
    --------
    ...

    """
```

Then create a vtk compatible faces array.  I wouldn't use `np.ones(...)*` as there's an additional multiplication step there.  Probably just construct an empty array, fill the first column of the array with the number of points per face, then populate the remainder with the face connectivity.

**Comment by sampotter** [comment]
> > It would be good if PyVista threw an exception under these circumstances.
> 
> Agreed. VTK segfaults due to memory access issues, and it would be great to check the validity of the array prior to sending it to VTK. However, the tradeoff would be that it would take longer to construct the `PolyData` as error checking it in numpy or pure python wouldn't be fast. We could consider adding a `check_validity` parameter to the `PolyData` constructor.
> 

I agree with that, although it's not completely clear to me what bug exactly is occurring. Although, I haven't tried figuring this out, either.

If a flag is added, I think a default value of `True` would make the most sense---if people really need the extra speed, _then_ it could be turned off. But it seems like adding a safeguard which would prevent Python from crashing (possibly at the very end of a large simulation!) would be beneficial.

> > If it doesn't already have one, it would be very helpful if PyVista had a convenience method for building a PolyData from a vertex and face array pair, since this is such a common use case.
> 
> It doesn't, but we could add one fairly easily. Would you mind submitting a PR that adds one to at `pyvista/utilities/helpers.py`? Perhaps something like:
> 
> ```python
> def construct_mesh(points, faces):
>    """Create a mesh given a square face connectivity array without padding.
> 
>     Examples
>     --------
>     ...
> 
>     """
> ```
> 
> Then create a vtk compatible faces array. I wouldn't use `np.ones(...)*` as there's an additional multiplication step there. Probably just construct an empty array, fill the first column of the array with the number of points per face, then populate the remainder with the face connectivity.

Sure. I went ahead and did so. Please see https://github.com/pyvista/pyvista/pull/1101.

**Comment by akaszynski** [comment]
> I agree with that, although it's not completely clear to me what bug exactly is occurring. Although, I haven't tried figuring this out, either.

The simpliest example I can think of is this:
```
[0, 1, 2, 3, 4, 5]
```
Imagine this is a flattened face array containing the faces `[0, 1, 2]` and `[3, 4, 5]`.  With your convenience function, we can easily construct a mesh with this.  However, VTK meshes can have faces with any number of points, so it uses the first value of the face to denote the size.  So, the flat face array would be: `[3, 0, 1, 2, 3, 3, 4, 5]`.  This lets VTK "walk" through the array, figuring out the sizes of faces along the way.

In the toy example however, VTK sees the following isolated faces `[0]` (null face), `[1, 2]` (a single point), and `[3, 4, 5]` (a triangle with a missing face index).  As this is C and there's no bounds checking implemented in VTK, as it's walking through the face indices on the final face, it will read 4, 5, and then expect a final value (which doesn't exist).  It will then read some value in memory which might be negative or very large, leading the program to read some other area in the memory, probably outside of the Python allocated memory, and then segfault.

Hope that makes sense.

**Comment by sampotter** [comment]
Ah, okay. Figured this was what was going on. Thanks for the detailed explanation.

I agree, writing an algorithm to quickly validate the VTK-style index list in vanilla Python would be unlikely to run very quickly. It would be easy enough to write something to check this using Cython, but it doesn't appear that speeding things up using Cython is a part of the PyVista way of doing things, which is reasonable.

As an alternative, I took a look at what's available in VTK (I am _not_ a VTK expert, although interested in picking it up), and it looks like `vtkCellArray` has an `IsValid` function which _might_ be what we're looking for, although it appears that the format of a `vtkCellArray` may be a bit different from what is passed to `pv.PolyData` (i.e., it involves offsets instead of face counts).

If this function fits the bill, maybe it would be straightforward to just call `vtkCellArray::IsValid` using the Python VTK interface when constructing `pv.PolyData` if your proposed `check_validity` keyword argument is `True`.

**Comment by akaszynski** [comment]
> but it doesn't appear that speeding things up using Cython is a part of the PyVista way of doing things, which is reasonable.

Actually, I'd be happy to incorporate cython into `PyVista`.  I use it quite a bit with other python projects either as a bridge between pure C/C++ or to speed up semi-trivial functions (like this one).  I'd appreciate a PR containing your cython code.  To separate out the module, create a directory `pyvista/cython` and we can place our `pyx` files there.

**Comment by sampotter** [comment]
Unfortunately, I don't have time to set up PyVista for use with Cython right now.

I spent a little more time thinking about this, and I don't think the current API for pv.PolyData makes much sense. From what I can tell, there are two modes:

1. Pass faces as a 1D array with each tuple of face indices preceded by the tuple length.
2. Pass a 2D array, where the first column contains the number of face indices in each row.

One problem with 2 is that since the has a fixed number of columns, the first row should always just be cells.shape[1] - 1. So, this is redundant information, and it doesn't really make any sense to pass it. It would be more sensible to change 2 so that when an `N x P` array is passed, a column containing P is blocked together with this array (resulting in an array with shape `(N, P + 1)`) before passing it on to `pv.CellArray` (and after looking at `pv.CellArray` for a bit, this class should probably take this approach, as well).

Thoughts? My worry here is that there would be some breaking changes. But IMO this change would make the API a lot more intuitive.

**Comment by user27182** [comment]
The mesh validation methods in #8164 can help address this. With the current draft API:
``` py
import pyvista as pv
poly = pv.PolyData([0.0,0.0,0.0],faces=[3, 0, 1, 2])
poly.validate_mesh(action='error')
# pyvista.core.errors.InvalidMeshError: PolyData mesh is not valid due to the following problems:
#  - Mesh has 1 cell with invalid point references. Invalid cell id: [0]
```

**Comment by user27182** [comment]
Many separate issues were raised here. The original issue with this example crashing python:

``` py
import meshzoo
import numpy as np
import pyvista as pv
V, F = meshzoo.triangle(8)
V = V.T
WRONG_NUMBER = 1
pv.PolyData(
     V,
     np.concatenate([
         WRONG_NUMBER*np.ones((F.shape[0], 1), dtype=F.dtype),
         F
     ], axis=1)
)
```

has been fixed. If we try this code, we get this error:
```
pyvista.core.errors.CellSizeError: `faces` cell array size is invalid.
```

> If it doesn't already have one, it would be very helpful if PyVista had a convenience method for building a PolyData from a vertex and face array pair, since this is such a common use case.

There is now an API for this:
https://docs.pyvista.org/api/core/_autosummary/pyvista.polydata.from_regular_faces

And in terms of adding a dedicated "check validity" keyword to enable this, the `CellSizeError` is _already_ enabled by default, and additional optional validation can now be done with [validate_mesh](https://dev.pyvista.org/api/core/_autosummary/pyvista.dataobjectfilters.validate_mesh)

And so I think this can be closed. If anything mentioned in this issue has been missed, please create a new issue with details.
